examples/field_examples/viking_graben_line12/fwi/fwi.jl

```diff
@@ -0,0 +1,278 @@
+
+grad_mem = 40 # Based on n and CFL condition
+
+mem = Sys.free_memory()/(1024^3)
+t_sub = max(1, ceil(Int, nworkers()*grad_mem/mem))
+@info "subsampling factor: $t_sub"
+
+############# FWI #############
+# JUDI options
+global jopt = JUDI.Options(
+    IC = "fwi",
+    limit_m = true,
+    buffer_size = buffer_size,
+    optimal_checkpointing=false,
+    # subsampling_factor=t_sub,     # subsampling_factor with space_order set leads to exception (probably bug)
+    free_surface=true,  # free_surface is on to model multiples as well
+    space_order=16)     # increase space order for > 12 Hz source wavelet
+
+global modelCell = Vector{Array{Float32}}(undef, 0)
+global gradientCell = Vector{Array{Float32}}(undef, 0)
+
+# optimization parameters
+niterations = 10
+shot_from = 1
+shot_to = length(d_obs)
+shot_step = 3   # we may want to calculate only each Nth shot to economy time
+count = 0
+fhistory = Vector{Float32}(undef, 0)
+mute_reflections = false
+mute_turning = false
+
+function save_results(x,z,m,grad,fhistory,v_plot_title,grad_plot_title,v_file,grad_file,h5_file)
+    @info "save_results: $h5_file"
+    n = (length(x),length(z))
+    o = (x[1],z[1])
+    d = (x[2]-x[1],z[2]-z[1])
+    fwi_plt = Plots.heatmap(x, z, sqrt.(1f0 ./ adjoint(reshape(m,n))), c=:rainbow, 
+        xlims=(x[1],x[end]), 
+        ylims=(z[1],z[end]), yflip=true,
+        title=v_plot_title,
+        xlabel="Lateral position [km]",
+        ylabel="Depth [km]",
+        dpi=600)
+    Plots.savefig(fwi_plt, v_file)
+
+    grad_plt = Plots.heatmap(x, z, adjoint(reshape(grad,n)), c=:bluesreds, 
+        xlims=(x[1],x[end]), 
+        ylims=(z[1],z[end]), yflip=true,
+        clim=(-maximum(grad)/5f0, maximum(grad)/5f0),
+        title=grad_plot_title,
+        xlabel="Lateral position [km]",
+        ylabel="Depth [km]",
+        dpi=600)
+    Plots.savefig(grad_plt,grad_file)
+
+    h5open(h5_file, "w") do file
+        write(file, 
+            "v", sqrt.(1f0 ./ reshape(m, n)), 
+            "m", reshape(m, n), 
+            "o", collect(o.*1000f0), 
+            "n", collect(n), 
+            "d", collect(d.*1000f0),
+            "fhistory", fhistory,
+            "grad", reshape(grad, n))
+    end
+end
+
+# NLopt objective function
+function nlopt_obj_fun!(m_update, grad)
+
+    global x, z, count, jopt, gradientCell, seabed_ind;
+    count += 1
+
+    # Update model
+    model0.m .= Float32.(reshape(m_update, model0.n))
+    if modeling_type == "acoustic"
+        model0.b .= Float32.(reshape(1f0./rho_from_slowness(model0.m), model0.n))
+        model0.b[:,1:seabed_ind] .= 1f0/rhowater
+    end
+
+    # Select batch and calculate gradient
+    # Subsampling the number of sources should in practice never be used for second order methods such as L-BFGS.
+    indsrc = shot_from:shot_step:shot_to
+    if mute_reflections
+        fval, gradient = fwi_objective(model0, q[indsrc], Ml_tur[indsrc]*d_obs[indsrc], options=jopt)
```

Comment:
for filtering you would just have an extra `Ml_freq[indsrc]` in front of `q` and `d_obs`